trajopt_ifopt/src/cartesian_line_constraint.cpp

```diff
@@ -0,0 +1,239 @@
+{
+CartLineConstraint::CartLineConstraint(const Eigen::Isometry3d& origin_pose,
+                                      const Eigen::Isometry3d& target_pose,
+                                      CartPosKinematicInfo::ConstPtr kinematic_info,
+                                      JointPosition::ConstPtr position_var,
+                                      const std::string& name)
+  : ifopt::ConstraintSet(6, name)
+  , position_var_(std::move(position_var))
+  , line_start_(origin_pose)
+  , line_end_(target_pose)
+  , kinematic_info_(std::move(kinematic_info))
+{
+  // Set the n_dof and n_vars for convenience
+  n_dof_ = kinematic_info_->manip->numJoints();
+  assert(n_dof_ > 0);
+
+  bounds_ = std::vector<ifopt::Bounds>(6, ifopt::BoundZero);
+
+  // calculate the equation of the constraint line
+  line_  = line_end_.translation() - line_start_.translation();//<< line_end_.translation().x() - line_start_.translation().x(), line_end_.translation().y() - line_start_.translation().y(), line_end_.translation().z() - line_start_.translation().z();
+
+}
+
+Eigen::VectorXd CartLineConstraint::CalcValues(const Eigen::Ref<const Eigen::VectorXd>& joint_vals)
+{
+
+  Eigen::Isometry3d new_pose;
+  kinematic_info_->manip->calcFwdKin(new_pose, joint_vals, kinematic_info_->kin_link->link_name);
+
+  new_pose = kinematic_info_->world_to_base * new_pose * kinematic_info_->kin_link->transform * kinematic_info_->tcp;
+
+  // For Jacobian Calc, we need the inverse of the nearest point, D, to new Pose, C, on the constraint line AB
+  // This will not honor tcp orientation. Not used bc this returns a scalar error, need vector/direction components
+//  Eigen::Vector3d d1,d2, pose_err;
+
+  //AC, new pose to first point on line
+//  d1 = new_pose.translation() - line_start_.translation();
+  //BC, new pose to second point on line
+//  d2 = new_pose.translation() - line_end_.translation();
+  //Distance = |AC * BC| / |AB|
+
+//  pose_err = d1.cross(d2).squaredNorm() /  line_.squaredNorm();
+
+  //Find D, the closest point on the line to C, the new point
+
+  //AC, new pose to first point on line
+  Eigen::Vector3d d1 = new_pose.translation() - line_start_.translation();
+
+  // ((AB * AC ) / (|AB|^2)) * (AB)
+  Eigen::Isometry3d temp_ = Eigen::Isometry3d::Identity();
+  double line_point_dist_ = (line_.dot(d1) / line_.dot(line_));
+  if (line_point_dist_ > 1.0)
+  {
+      temp_.translate(line_end_.translation());
+  }
+  else if (line_point_dist_ < 0)
+  {
+      temp_.translate(line_start_.translation());
+  }
+  else
+  {
+      temp_.translate(line_point_dist_ * line_);
+  }
+  line_point_ = temp_;
+  //SetLinePoint(temp_);
+
+  //pose error is the vector from new pose C to nearest point on line AB, D
+  Eigen::Vector3d pose_err = line_point_.translation() - new_pose.translation();
+
+  // @TODO: Handle orientation
+  Eigen::Vector4d rot_ = Eigen::Vector4d::Identity();
+  Eigen::VectorXd err = concat(pose_err, rot_); //calcRotationalError(pose_err.rotation()));
```

Comment:
The comments in the header say err should be 6 long. This will be 7 won't it? Also, orientation is unimplemented, shouldn't rot be zeros instead of ones? It probably won't effect the optimization, but it will make the error look bigger than it actually is.